src/v/archival/ntp_archiver_service.cc

```diff
@@ -295,28 +295,68 @@ ss::future<cloud_storage::download_result> ntp_archiver::sync_manifest() {
-        auto offset
-          = _partition->archival_meta_stm()->manifest().get_last_offset()
-            + model::offset(1);
-        // TODO: this code needs to be updated when the compacted segment
-        // uploads are merged.
+        // Several things has to be done:
+        // - Add all segments between old last_offset and new last_offset
+        // - Compare all segments below last compacted offset with their
+        //   counterparts in the old manifest and re-add them if they are
+        //   diferent.
+        // - Apply new start_offset if it's different
+        auto offset = model::next_offset(manifest().get_last_offset());
         for (auto it = m.segment_containing(offset); it != m.end(); it++) {
             mdiff.push_back(it->second);
         }
+
+        bool needs_cleanup = false;
+        auto old_start_offset = manifest().get_start_offset();
+        auto new_start_offset = m.get_start_offset();
+        for (const auto& s : m) {
+            if (
+              s.second.committed_offset
+                <= m.get_last_uploaded_compacted_offset()
+              && s.second.base_offset >= new_start_offset) {
+                auto iter = manifest().get(s.first);
+                if (iter && *iter != s.second) {
+                    mdiff.push_back(s.second);
+                    needs_cleanup = true;
+                }
+            } else {
+                break;
+            }
+        }
+
         auto sync_timeout = config::shard_local_cfg()
                               .cloud_storage_metadata_sync_timeout_ms.value();
         auto deadline = ss::lowres_clock::now() + sync_timeout;
-        auto error = co_await _partition->archival_meta_stm()->add_segments(
-          mdiff, deadline, _as);
-        if (
-          error != cluster::errc::success
-          && error != cluster::errc::not_leader) {
+        // The commands to update the manifest need to be batched together,
+        // otherwise the read-replica will be able to see partial update. Also,
+        // the batching is more efficient.
+        auto builder = _partition->archival_meta_stm()->batch_start(
+          deadline, _as);
+        try {
+            builder.add_segments(std::move(mdiff));
+            if (
+              new_start_offset.has_value()
+              && old_start_offset.value_or(model::offset())
+                   != new_start_offset.value()) {
+                builder.truncate(new_start_offset.value());
+                needs_cleanup = true;
+            }
+            if (needs_cleanup) {
+                // We only need to replicate this command if the
+                // manifest will be truncated or compacted segments
+                // will be added by previous commands.
+                builder.cleanup_metadata();
+            }
+            co_await builder.replicate();
+            co_return cloud_storage::download_result::success;
+        } catch (...) {
+            builder.discard();
```

Comment:
nit: doing discard here is optional as `_completed` is set first thing in the `replicate()` method, right?